### PR TITLE
DOC: improve documentation of canonical form and set_precision

### LINKED ISSUE #1950: set_precision() enforces clockwise order on coordinates
The following code shows the behaviour:
```
import shapely.geometry as geom
import shapely.ops

cw_poly = geom.Polygon([(0, 0), (0, 1), (1, 1), (0, 0)])
ccw_poly = cw_poly.reverse()

print(f"""
{geom.LinearRing(cw_poly.exterior.coords).is_ccw=}
{geom.LinearRing(ccw_poly.exterior.coords).is_ccw=}
{geom.LinearRing(shapely.set_precision(cw_poly, 1e-4).exterior.coords).is_ccw=}
{geom.LinearRing(shapely.set_precision(ccw_poly, 1e-4).exterior.coords).is_ccw=}""")
```
giving the following result:
```
geom.LinearRing(cw_poly.exterior.coords).is_ccw=False
geom.LinearRing(ccw_poly.exterior.coords).is_ccw=True
geom.LinearRing(shapely.set_precision(cw_poly, 1e-4).exterior.coords).is_ccw=False
geom.LinearRing(shapely.set_precision(ccw_poly, 1e-4).exterior.coords).is_ccw=False
```
The last one is the problem. The `ccw_poly` is CCW ordered but the result returned by `shapely.set_precision()` is CW ordered. So after the polygon returned by `shapely.set_precision` is always clockwise ordered. 

That's not really a bug, it's a feature, but it threw me... and I can't see it spelled out anywhere in the manual. 

Some code I'm working on depended on the order, and the function I was using to enforce CCW order was applying set_precision before returning the polygon, hence invisibly ensuring that the returned polygon was not CCW but CW!

Maybe update the docs to reflect this information?

### LINKED ISSUE #1952: set_precision() changes order of coordinates
`shapely.set_precision()` as reported in issue #1950 has side-effects. Here's another one. The position of points in the coordinate sequence of a polygon is altered.
```
p1 = shapely.geometry.Polygon([(0, 0), (0, 1), (1, 1), (1, 0)])
p2 = shapely.set_precision(p1, 1e-3)
p1, p2
```
This gives us
```
(<POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))>, <POLYGON ((0 1, 1 1, 1 0, 0 0, 0 1))>)
```
The polygon `p1` is clockwise, as is `p2` but they start at different coordinates in the sequence!

As for the previous issue this isn't a bug _per se_ but it's poorly documented, and... well... it's an unexpected side-effect of simply setting the precision.

### PR DESCRIPTION
**Pull request by theroggy**
Improvements to `set_precision` documentation:
- it changes the order of points,... which doesn't seem to be expected to happen. So better to mention this explicitly in its documentation.
- some more clarification on what `set_precision` does, e.g. the collapsing of narrow parts of polygons.

The "canonical form" also isn't documented in detail... so this PR tries to solve this as well.

Resolves #1950 
Resolves #1952

### CODE REVIEW

File: docs/geometry.rst [hunk 1]
```diff
@@ -173,3 +173,17 @@ specified, rounding precision will be disabled showing full precision.
...
+
+.. _canonical-form:
+
+Canonical form
+--------------
+When geometries are constructed, they are created according to some conventions:
+
+* the coordinates of exterior rings follow a clockwise orientation, interior rings
+  have a counter-clockwise orientation. This is the opposite of the OGC specifications
+  because the choise was made before this was included in the standard.
```
Comment (DOSull) [inline docs/geometry.rst:185]: choise -> choice

File: docs/geometry.rst [hunk 2]
```diff
@@ -173,3 +173,17 @@ specified, rounding precision will be disabled showing full precision.
 Format types ``'x'`` and ``'X'`` show a hex-encoded string representation of
 WKB or Well-Known Binary, with the case of the output matched the
 case of the format type character.
+
+.. _canonical-form:
```
Comment (DOSull) [inline docs/geometry.rst:177]: I might be wrong about this, but I think canonical form may also impose some convention as to which point in a polygon comes first? Or perhaps it is just the shapely implementation that does this. Anyway, if so, this should also be noted and explained. 

As I have said, I might be wrong, but e.g. (note that p1 is in CW order) so this is not a result of a winding order reversal:
```
>>> import shapely
>>> p1 = shapely.geometry.Polygon([(0, 0), (0, 100), (100, 100), (100, 0)])
>>> p2 = shapely.set_precision(p1, grid_size = 1e-3)
>>> print(f"{p1=}\n{p2=}")
p1=<POLYGON ((0 0, 0 100, 100 100, 100 0, 0 0))>
p2=<POLYGON ((0 100, 100 100, 100 0, 0 0, 0 100))>
```
This behaviour might be less likely to bother people, but it's still a behaviour!

**Review by DOSull — COMMENTED** [review]
I've made a couple of inline comments - a spelling correction, but also a more substantive comment concerning the order of coordinates in canonical form, which may also require explanation.

File: docs/geometry.rst (hunk 2 above)
Comment (theroggy) [reply docs/geometry.rst:177]: @DOSull the second point is the following:
* the starting point of rings is lower left.

This isn't clear?

Comment (theroggy) [reply docs/geometry.rst:177]: Hmm... good point.

I did some extra tests... and apparently the result of `set_precision` isn't completely in canonical form... that specific aspect is not correcly applied...

Not sure what to do with it... Strictly speaking it is probably a bug in geos... but....

```
import shapely

p1 = shapely.geometry.Polygon([(0, 0), (100, 0), (100, 100), (0, 100)])
p2 = shapely.set_precision(p1, grid_size=1e-3)
p3 = shapely.normalize(p1)
p4 = shapely.normalize(p2)

print(f"{p1=} (counter-clockwise)")
print(f"{p2=} (clockwise)")
print(f"{p3=} (clockwise)")
print(f"{p4=} (clockwise)")
```

Output:
```
p1=<POLYGON ((0 0, 100 0, 100 100, 0 100, 0 0))> (counter-clockwise)
p2=<POLYGON ((0 100, 100 100, 100 0, 0 0, 0 100))> (clockwise)
p3=<POLYGON ((0 0, 0 100, 100 100, 100 0, 0 0))> (clockwise)
p4=<POLYGON ((0 0, 0 100, 100 100, 100 0, 0 0))> (clockwise)
```

Comment (theroggy) [reply docs/geometry.rst:177]: @DOSull I did some more tests and updated the documentation accordingly.

**Comment by theroggy** [comment]
> Seems OK to me FWIW - although this does seem to point to a need for some more consistency in when canonical forms are applied

To check if it is to expected that the geometry is returned normalized I opened an issue in geos https://github.com/libgeos/geos/issues/1032 ... please have a look...

File: docs/geometry.rst [hunk 3]
```diff
@@ -173,3 +173,31 @@ specified, rounding precision will be disabled showing full precision.
...
+  undefined
+
+When :func:`~shapely.normalize` is used, the "strict" canonical form is applied. This
+type of normalization is meant to be stable, so changes to it will be avoided if
+possible:
+
+- the coordinates of exterior rings follow a clockwise orientation and interior
+  rings have a counter-clockwise orientation
+- the starting point of rings is lower left
+- elements in collections are ordered by geometry type
```
Comment (jorisvandenbossche) [inline docs/geometry.rst:201]: And then in you have multiple ones of the same type? Or does that get "normalized" in a Multi type?

File: docs/geometry.rst [hunk 4]
```diff
@@ -173,3 +173,31 @@ specified, rounding precision will be disabled showing full precision.
...
+When operations are applied on geometries the result is returned according to some
+conventions.
+
+In most cases, geometries will be returned in "mild" canonical form. There is no goal
+to keep this form stable, so it is expected to change in future versions of GEOS:
+
+- the coordinates of exterior rings follow a clockwise orientation and interior
+  rings have a counter-clockwise orientation. This is the opposite of the OGC
+  specifications because the choice was made before this was included in the standard.
+- the starting point of rings can be changed, but it is undefined
```
Comment (jorisvandenbossche) [inline docs/geometry.rst:190]:
```suggestion
- the starting point of rings can be changed in the output, but the exact order is undefined and should not be relied upon
```

File: shapely/_geometry.py [hunk 5]
```diff
@@ -731,20 +731,23 @@ def set_precision(geometry, grid_size, mode="valid_output", **kwargs):
...
-
-    Note: subsequent operations will always be performed in the precision of
-    the geometry with higher precision (smaller "grid_size"). That same
-    precision will be attached to the operation outputs.
-
-    Also note: input geometries should be geometrically valid; unexpected
-    results may occur if input geometries are not.
-
-    Returns None if geometry is None.
+    Coordinates will be rounded if the precision grid specified is less precise than the
```
Comment (jorisvandenbossche) [inline shapely/_geometry.py:734]: Can you keep this to the original line width? (I know in general the project allows 88 per the formatter, but for docstrings this is 1) not actually enforced by the linter, and 2) generally the docstrings are wrapped below 80 (which IMO improves readability))

File: shapely/_geometry.py [hunk 6]
```diff
@@ -731,20 +731,23 @@ def set_precision(geometry, grid_size, mode="valid_output", **kwargs):
...
+    multipolygons or empty geometries. Z values, if present, will not be modified.
+
+    Notes:
+
+    * subsequent operations will always be performed in the precision of
+      the geometry with higher precision (smaller "grid_size"). That same
+      precision will be attached to the operation outputs.
+    * input geometries should be geometrically valid; unexpected results may
+      occur if input geometries are not.
+    * the geometry returned will be in :ref:`mild canonical form <canonical-form>`.
```
Comment (jorisvandenbossche) [inline shapely/_geometry.py:749]:
```suggestion
    * the geometry returned will be in :ref:`mild canonical form <canonical-form>`,
      and the order of vertices can change and should not be relied upon.
```

(I think that's the main gotcha that triggered some issue? So in addition to putting the link, summarizing it like that might be helpful)

**Review by jorisvandenbossche — COMMENTED** [review]
This for writing this up! Finally did a read through and added a few comments

File: shapely/_geometry.py (hunk 5 above)
Comment (theroggy) [reply shapely/_geometry.py:734]: ok

File: docs/geometry.rst (hunk 3 above)
Comment (theroggy) [reply docs/geometry.rst:201]: Indeed the same from the same type. It seems that those are ordered from top to bottom and from right to left.

``` py
import shapely

coll = shapely.GeometryCollection(
    [
        shapely.box(0, 0, 10, 10),
        shapely.Point(50, 50),
        shapely.MultiPoint([(50, 50), (55, 55)]),
        shapely.LineString([(60, 60), (70, 70)]),
        shapely.box(-20, -20, -30, -30),
        shapely.box(20, 20, 30, 30),
        shapely.box(-30, 20, -20, 30),
        shapely.MultiLineString([[(60, 60), (70, 70)]]),
        shapely.box(20, -20, 30, -30),
        shapely.MultiLineString([[(60, 60), (70, 70)], [(80, 80), (90, 90)]]),
    ]
)
print(f"{str(coll)=}")
print(f"{str(coll.normalize())=}")

coll2 = shapely.GeometryCollection(
    [
        shapely.box(-30, 20, -20, 30),
        shapely.box(20, -20, 30, -30),
        shapely.box(20, 20, 30, 30),
        shapely.LineString([(60, 60), (70, 70)]),
        shapely.Point(50, 50),
        shapely.MultiPoint([(50, 50), (55, 55)]),
        shapely.box(0, 0, 10, 10),
        shapely.box(-20, -20, -30, -30),
        shapely.MultiLineString([[(60, 60), (70, 70)]]),
        shapely.MultiLineString([[(60, 60), (70, 70)], [(80, 80), (90, 90)]]),
    ]
)
print(f"{str(coll2)=}")
print(f"{str(coll2.normalize())=}")
```
Result:
```
str(coll)='GEOMETRYCOLLECTION (POLYGON ((10 0, 10 10, 0 10, 0 0, 10 0)), POINT (50 50), MULTIPOINT ((50 50), (55 55)), LINESTRING (60 60, 70 70), POLYGON ((-30 -20, -30 -30, -20 -30, -20 -20, -30 -20)), POLYGON ((30 20, 30 30, 20 30, 20 20, 30 20)), POLYGON ((-20 20, -20 30, -30 30, -30 20, -20 20)), MULTILINESTRING ((60 60, 70 70)), POLYGON ((30 -20, 30 -30, 20 -30, 20 -20, 30 -20)), MULTILINESTRING ((60 60, 70 70), (80 80, 90 90)))'
str(coll.normalize())='GEOMETRYCOLLECTION (POLYGON ((20 20, 20 30, 30 30, 30 20, 20 20)), POLYGON ((20 -30, 20 -20, 30 -20, 30 -30, 20 -30)), POLYGON ((0 0, 0 10, 10 10, 10 0, 0 0)), POLYGON ((-30 20, -30 30, -20 30, -20 20, -30 20)), POLYGON ((-30 -30, -30 -20, -20 -20, -20 -30, -30 -30)), MULTILINESTRING ((80 80, 90 90), (60 60, 70 70)), MULTILINESTRING ((60 60, 70 70)), LINESTRING (60 60, 70 70), MULTIPOINT ((55 55), (50 50)), POINT (50 50))'
str(coll2)='GEOMETRYCOLLECTION (POLYGON ((-20 20, -20 30, -30 30, -30 20, -20 20)), POLYGON ((30 -20, 30 -30, 20 -30, 20 -20, 30 -20)), POLYGON ((30 20, 30 30, 20 30, 20 20, 30 20)), LINESTRING (60 60, 70 70), POINT (50 50), MULTIPOINT ((50 50), (55 55)), POLYGON ((10 0, 10 10, 0 10, 0 0, 10 0)), POLYGON ((-30 -20, -30 -30, -20 -30, -20 -20, -30 -20)), MULTILINESTRING ((60 60, 70 70)), MULTILINESTRING ((60 60, 70 70), (80 80, 90 90)))'
str(coll2.normalize())='GEOMETRYCOLLECTION (POLYGON ((20 20, 20 30, 30 30, 30 20, 20 20)), POLYGON ((20 -30, 20 -20, 30 -20, 30 -30, 20 -30)), POLYGON ((0 0, 0 10, 10 10, 10 0, 0 0)), POLYGON ((-30 20, -30 30, -20 30, -20 20, -30 20)), POLYGON ((-30 -30, -30 -20, -20 -20, -20 -30, -30 -30)), MULTILINESTRING ((80 80, 90 90), (60 60, 70 70)), MULTILINESTRING ((60 60, 70 70)), LINESTRING (60 60, 70 70), MULTIPOINT ((55 55), (50 50)), POINT (50 50))'
```

**Review by jorisvandenbossche — APPROVED** [review]
Thanks again!

**Pull request merged**